File: dotnet/Lexy.Poc/Lexy.Poc.Core/Language/Expression.cs
using System.IO;
using Lexy.Poc.Core.Parser;
using Lexy.Poc.Core.Parser.Tokens;

namespace Lexy.Poc.Core.Language
{
    public class Expression
    {
        public Line SourceLine { get; }
        public Token[] Tokens { get; }

        private Expression(Line sourceLine, Token[] tokens)
        {
            SourceLine = sourceLine;
            Tokens = tokens;
        }

        public static Expression Parse(Line sourceLine, Token[] tokens)
        {
            var expression = new Expression(sourceLine, tokens);
            return expression;
        }

        public override string ToString()
        {
            var writer = new StringWriter();
            foreach (var token in Tokens)
            {
                writer.Write(token.Value);
            }
            return writer.ToString();
        }
    }
}

File: typescipt/existing/src/language/expressions/expression.ts





namespace Lexy.Compiler.Language.Expressions;

public abstract class Expression : Node
{
   public ExpressionSource Source { get; }

   protected Expression(ExpressionSource source, SourceReference reference) : base(reference)
   {
     Source = source ?? throw new ArgumentNullException(nameof(source));
   }

   public override string ToString()
   {
     var writer = new StringWriter();
     foreach (var token in Source.Tokens) writer.Write(token.Value);
     return writer.ToString();
   }

   public abstract VariableType DeriveType(IValidationContext context);
}
